app/uk/gov/hmrc/gform/obligation/ObligationService.scala

```diff
@@ -17,30 +17,114 @@
 package uk.gov.hmrc.gform.obligation
 import java.text.SimpleDateFormat
 
+import uk.gov.hmrc.gform.auth.AuthService
+import uk.gov.hmrc.gform.auth.models.MaterialisedRetrievals
+import uk.gov.hmrc.gform.controllers.helpers.FormDataHelpers
 import uk.gov.hmrc.gform.gformbackend.GformConnector
 import uk.gov.hmrc.gform.sharedmodel._
+import uk.gov.hmrc.gform.sharedmodel.form.{ Form, FormId, UserData }
 import uk.gov.hmrc.gform.sharedmodel.formtemplate._
 import uk.gov.hmrc.http.HeaderCarrier
-import scala.concurrent.ExecutionContext
+
+import scala.concurrent.{ ExecutionContext, Future }
 
 class ObligationService(gformConnector: GformConnector) {
 
   val stringToDate = new SimpleDateFormat("yyyy-MM-dd")
 
-  def lookupObligationsMultiple(formTemplate: FormTemplate)(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
+  def lookupObligationsMultiple(
+    formTemplate: FormTemplate,
+    authService: AuthService,
+    retrievals: MaterialisedRetrievals,
+    form: Form)(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
+    val hmrcTaxPeriodIdentifiers = formTemplate.expandFormTemplateFull.allFCs.collect {
+      case IsHmrcTaxPeriod(el) => el
+    }
+    for {
+      listOfIdNumbers <- Future.sequence(
+                          hmrcTaxPeriodIdentifiers.map(
+                            i =>
+                              authService.evaluateSubmissionReference(
+                                i.idNumber,
+                                retrievals,
+                                formTemplate,
+                                FormDataHelpers.formDataMap(form.formData),
+                                form.envelopeId)))
+      replaceIdNumber <- Future(listOfIdNumbers.zip(hmrcTaxPeriodIdentifiers))
+      output <- gformConnector
+                 .getAllTaxPeriods(replaceIdNumber.map(a =>
+                   HmrcTaxPeriod(a._2.idType, TextExpression(Constant(a._1)), a._2.regimeType)))
```

Comment:
Can we name what is `a._1` and `a._2`?

Reply:
I'm not to sure what you mean. They were zipped just above this and naming them by making them different variables would defeat the point of doing that

Reply:
In couple of days as code grows, refactoring happen etc.
> They were zipped just above this ...
may become completely different part of code base so reader will have no clue whatsoever what `replaceIdNumber` represent.

It would be much nicer if mapping over `replaceIdNumber` happen straight after zipping.